src/Storages/MergeTree/ReplicatedMergeTreeCurrentlyRestoringFromBackup.cpp

```diff
@@ -0,0 +1,1126 @@
+    /// like "mutations/mutation-placeholder-0000000321".
+    std::vector<Int64> extractMutationNumbersFromPaths(const TemporaryZookeeperNodes & mutation_number_nodes) const
+    {
+        auto mutations_path = fs::path(storage.zookeeper_path) / "mutations";
+        auto prefix = mutations_path / kMutationPlaceholderPrefix;
+        return extractNumbersFromSequentialNodePaths(mutation_number_nodes.getPaths(), prefix);
+    }
+
+    /// Extracts numbers from the endings of multiple sequential node paths.
+    std::vector<Int64> extractNumbersFromSequentialNodePaths(const Strings & zookeeper_paths, const String & prefix) const
+    {
+        std::vector<Int64> res;
+        res.reserve(zookeeper_paths.size());
+        for (const String & zookeeper_path : zookeeper_paths)
+            res.emplace_back(extractNumberFromSequentialNodePath(zookeeper_path, prefix));
+        return res;
+    }
+
+    /// Extracts a number from the endings of a sequential node path.
+    Int64 extractNumberFromSequentialNodePath(const String & zookeeper_path, const String & prefix) const
+    {
+        if (!zookeeper_path.starts_with(prefix))
+        {
+            throw Exception(ErrorCodes::UNEXPECTED_NODE_IN_ZOOKEEPER, "Sequential node '{}' was created without prefix '{}'",
+                            zookeeper_path, prefix);
+        }
+        return parseFromString<Int64>(zookeeper_path.substr(prefix.length()));
+    }
+
+    /// Tries to extract a number from the endings of a sequential node path without throwing exceptions.
+    std::optional<Int64> tryExtractNumberFromSequentialNodePath(const String & zookeeper_path, const String & prefix) const
+    {
+        if (!zookeeper_path.starts_with(prefix))
+            return {};
+
+        Int64 res;
+        if (!tryParse(res, zookeeper_path.substr(prefix.length())))
+            return {};
+
+        return res;
+    }
+
+    /// Checks that there no parts exist in the table.
+    /// The function also checks that no parts will be inserted / attached soon due to a concurrent process.
+    void checkNoPartsExist(const ZooKeeperWithFaultInjectionPtr & zookeeper, const Strings & ignore_block_number_zk_paths, CheckForNoPartsReason reason) const
+    {
+        if (reason == CheckForNoPartsReason::NONE)
+            return;
+
+        LOG_INFO(log, "Checking that no parts exist in the table before restoring its data (reason: {})", magic_enum::enum_name(reason));
+
+        /// The order of checks is important:
+        /// an INSERT command first allocates a block number, then it creates a log entry, and finally it replicates the part into all replicas.
+        /// That's why it's more reliable to check first block numbers, then the replication queue and finally the current replica itself.
+
+        auto allocated_block_number = findAnyAllocatedBlockNumber(zookeeper, ignore_block_number_zk_paths);
```

Comment:
This comment is still confusing. Also, it's possible that a new entry was added to the replication log just after we pulled the log in `findAnyPartInReplicationQueue` (yes, it's not a problem because we call checkNoPartsExist twice, but the order does not protect from anything)